src/actions/pause/mapScopes.js

```diff
@@ -85,6 +86,65 @@ export function mapScopes(scopes: Promise<Scope>, frame: Frame) {
   };
 }
 
+function batchScopeMappings(
+  originalAstScopes: Array<SourceScope>,
+  source: Source,
+  sourceMaps: any
+) {
+  const precalculatedRanges = new Map();
+  const precalculatedLocations = new Map();
+
+  // Explicitly dispatch all of the sourcemap requests synchronously up front so
+  // that they will be batched into a single request for the worker to process.
+  for (const item of originalAstScopes) {
+    for (const name of Object.keys(item.bindings)) {
+      for (const ref of item.bindings[name].refs) {
+        const locs = [ref];
+        if (ref.type === "decl") {
+          locs.push(ref.declaration);
+        }
+
+        for (const loc of locs) {
+          precalculatedRanges.set(
+            buildLocationKey(loc.start),
+            sourceMaps.getGeneratedRanges(loc.start, source)
+          );
+          precalculatedLocations.set(
+            buildLocationKey(loc.start),
+            sourceMaps.getGeneratedLocation(loc.start, source)
+          );
+          precalculatedLocations.set(
+            buildLocationKey(loc.end),
+            sourceMaps.getGeneratedLocation(loc.end, source)
+          );
+        }
+      }
+    }
+  }
+
+  return {
+    async getGeneratedRanges(pos, s) {
+      const key = buildLocationKey(pos);
+
+      if (s !== source || !precalculatedRanges.has(key)) {
+        throw new Error("Bad precalculated mapping");
```

Comment:
so... do we want to throw in firefox? if not we should use `utils/assert`

i guess we are catching this globally...

Reply:
Mostly this is to catch programming errors. I can't really imagine a case where this would throw in production but not during whatever work we're doing. If we did want to avoid this though, I'd probably say we update it to throw when `isDevelopment` and otherwise just call through to `sourceMaps` to do the query without batching.

Reply:
yeah, lets do that... small fix

Reply:
Updated!